crates/driver/src/tests/cases/protocol_fees.rs

```diff
@@ -111,6 +111,320 @@ async fn protocol_fee_test_case(test_case: TestCase) {
+            // The protocol fees are applied one by one to the orders, altering the order pricing
+            // For this example:
+            // -> First fee policy:
+            //  20 ETH surplus (50 ETH - 30 ETH), surplus policy: 50 %, fee 10 ETH
+            // -> Second fee policy:
+            // New sell amount in the Order: 40 ETH + 10 ETH (fee) = 50 ETH
+            // New surplus: 10 ETH, fee 5 ETH
+            // Total fee: 15 ETH
+            // Out of the 20 ETH of surplus, 15 ETH goes to fees and 5 ETH goes to the trader
+            solver: Amounts {
+                sell: 30.ether().into_wei(),
+                buy: 40.ether().into_wei(),
+            },
+            driver: Amounts {
+                sell: 45.ether().into_wei(),
+                buy: 40.ether().into_wei(),
+            },
+        },
+        expected_score: 20.ether().into_wei(),
+        fee_handler: FeeHandler::Driver,
+    };
+
+    protocol_fee_test_case(test_case).await;
+}
+
+#[tokio::test]
+#[ignore]
+async fn triple_surplus_protocol_fee_buy_order_not_capped() {
+    let fee_policy_surplus = Policy::Surplus {
+        factor: 0.5,
+        // high enough so we don't get capped by volume fee
+        max_volume_factor: 1.0,
+    };
+    let test_case = TestCase {
+        fee_policy: vec![
+            fee_policy_surplus.clone(),
+            fee_policy_surplus.clone(),
+            fee_policy_surplus,
+        ],
+        order: Order {
+            sell_amount: 50.ether().into_wei(),
+            buy_amount: 40.ether().into_wei(),
+            side: order::Side::Buy,
+        },
+        execution: Execution {
+            // 20 ETH surplus in sell token (after network fee)
+            // The protocol fees are applied one by one to the orders, altering the order pricing
+            // For this example:
+            // -> First fee policy:
+            //  20 ETH surplus (50 ETH - 30 ETH), surplus policy: 50 %, fee 10 ETH
+            // -> Second fee policy:
+            // New sell amount in the Order: 40 ETH + 10 ETH (fee) = 50 ETH
+            // New surplus: 10 ETH, fee 5 ETH
+            // -> Third fee policy:
+            // New sell amount in the Order: 50 ETH + 5 ETH (fee) = 55 ETH
+            // New surplus: 5 ETH, fee 2.5 ETH
+            // Total fee: 17.5 ETH
+            // Out of the 20 ETH of surplus, 17.5 ETH goes to fees and 2.5 ETH goes to the trader
+            solver: Amounts {
+                sell: 30.ether().into_wei(),
+                buy: 40.ether().into_wei(),
+            },
+            driver: Amounts {
+                sell: 47.5.ether().into_wei(),
+                buy: 40.ether().into_wei(),
+            },
+        },
+        expected_score: 20.ether().into_wei(),
+        fee_handler: FeeHandler::Driver,
+    };
+
+    protocol_fee_test_case(test_case).await;
+}
+
+#[tokio::test]
+#[ignore]
+async fn double_surplus_protocol_fee_sell_order_not_capped() {
```

Comment:
nit: I like the extended tests, but am afraid they might become a bit too large to maintain. Is there any value in having the double and triple test?

Reply:
@fleupold the double test and triple tests were really good for me to debug and catch the errors, since they are symmetrical. But they can be removed as they don't bring any extra value.

Reply:
I will leave the tests in the PR until @sunce86 reviews this PR, and then I will remove them if that is ok with you.

Reply:
Yes, we can probably remove double test if we already have triple test that covers it.